src/thread.rs

```diff
@@ -0,0 +1,483 @@
+//! Implements threads.
+
+use std::cell::RefCell;
+use std::convert::TryFrom;
+use std::num::NonZeroU32;
+
+use log::trace;
+
+use rustc_data_structures::fx::FxHashMap;
+use rustc_hir::def_id::DefId;
+use rustc_index::vec::{Idx, IndexVec};
+use rustc_middle::{
+    middle::codegen_fn_attrs::CodegenFnAttrFlags,
+    mir,
+    ty::{self, Instance},
+};
+
+use crate::*;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum SchedulingAction {
+    /// Execute step on the active thread.
+    ExecuteStep,
+    /// Execute destructors of the active thread.
+    ExecuteDtors,
+    /// Stop the program.
+    Stop,
+}
+
+/// A thread identifier.
+#[derive(Clone, Copy, Debug, PartialOrd, Ord, PartialEq, Eq, Hash)]
+pub struct ThreadId(usize);
+
+impl Idx for ThreadId {
+    fn new(idx: usize) -> Self {
+        ThreadId(idx)
+    }
+    fn index(self) -> usize {
+        self.0
+    }
+}
+
+impl From<u64> for ThreadId {
+    fn from(id: u64) -> Self {
+        Self(id as usize)
+    }
+}
+
+impl From<u32> for ThreadId {
+    fn from(id: u32) -> Self {
+        Self(id as usize)
+    }
+}
+
+impl ThreadId {
+    pub fn to_u32_scalar<'tcx>(&self) -> Scalar<Tag> {
+        Scalar::from_u32(u32::try_from(self.0).unwrap())
+    }
+}
+
+/// An identifier of a set of blocked threads.
+#[derive(Clone, Copy, Debug, PartialOrd, Ord, PartialEq, Eq, Hash)]
+pub struct BlockSetId(NonZeroU32);
+
+impl From<u32> for BlockSetId {
+    fn from(id: u32) -> Self {
+        Self(NonZeroU32::new(id).expect("0 is not a valid blockset id"))
+    }
+}
+
+impl BlockSetId {
+    pub fn to_u32_scalar<'tcx>(&self) -> Scalar<Tag> {
+        Scalar::from_u32(self.0.get())
+    }
+}
+
+/// The state of a thread.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum ThreadState {
+    /// The thread is enabled and can be executed.
+    Enabled,
+    /// The thread tried to join the specified thread and is blocked until that
+    /// thread terminates.
+    BlockedOnJoin(ThreadId),
+    /// The thread is blocked and belongs to the given blockset..
+    Blocked(BlockSetId),
+    /// The thread has terminated its execution (we do not delete terminated
+    /// threads.)
+    Terminated,
+}
+
+/// A thread.
+pub struct Thread<'mir, 'tcx> {
+    state: ThreadState,
+    /// Name of the thread.
+    thread_name: Option<Vec<u8>>,
```

Comment:
I couldn't find any code in libstd that *reads* the thread name from the system. Is this used for anything?
Also I seem to recall there was something special about the name of the main thread...?

Reply:
Ah it's just special in the sense that Rust libstd sets its name to `main` in `rt.rs`. Likely nothing Miri has to do there.

Reply:
> I couldn't find any code in libstd that reads the thread name from the system. Is this used for anything?

You can read this value by using `libc::prctl`. Not sure if real Rust code ever does that. However, the thread name is printed in the `Debug` [implementation](https://github.com/rust-lang/miri/pull/1284/files#diff-3e97dcab09523f9702269d7b101efbcbR132-R145) for `Thread`, which I think could be useful for debugging.